### ISSUE TITLE
Earmuff mode

### ISSUE THREAD

**Issue by InconsolableCellist** [Yellow-Dog-Man/Resonite-Issues#26]
### Is your feature request related to a problem? Please describe.

Like in most social VR games, it's possible to have so many people in a world/near you that the audio gets overwhelming.

### Describe the solution you'd like

I'd like something equivalent to VRChat's earmuff mode (an idea stolen from the mod version), so that you can drastically reduce the falloff distance and volume.

### Describe alternatives you've considered

You can individually lower people's volume in the session settings, but that's obviously not ideal and quite labor intensive, and you have to constantly adjust it based on who's near you.

**Comment by kulzae** [comment]
Isnt this covered by the whisper bubble? is there something that function lacks that this would provide?

**Comment by MegaShinySnivy** [comment]
I believe the point of this would be to be nonvisual, and to just reduce the volume of those around instead of muting them.

**Comment by InconsolableCellist** [comment]
I haven't found the whisper bubble yet, how can I use that? But yeah, this would either be non-visual or only show a visual cue during configuration, and then it'd fade away until you want to see it again.

**Comment by XDelta** [comment]
The whisper mode visual can be changed or hidden entirely if desired. You can switch to it as the second button in the top left of your dashboard. It is the purple mic icon
![ac60426e60](https://github.com/Yellow-Dog-Man/Resonite-Issues/assets/7883807/45c656cf-502c-4915-91d2-34fb99dba277)

**Comment by DoubleStyx** [comment]
@XDelta It's not possible to change the range of the whisper mode though iirc.

**Comment by InconsolableCellist** [comment]
I tried whisper mode but it has some big drawbacks. Features I'd consider necessary for an earmuff mode:

* Cone-based voice falloff: people not in a cone of vision should be quieter, with an adjustable soft or hard edge
* Outside and inside volume settings
* Distance and width of cone

**Comment by XDelta** [comment]
It does have settings for the volume outside of the bubble, with the inside volume being controlled by the voice setting as normal, in settings look for `Whisper Voice Volume (outside bubble)`

**Comment by XDelta** [comment]
Some worlds in game have implemented a more cone-like audio control system for 'facing attenuation'.

**Comment by XDelta** [comment]
> It's not possible to change the range of the whisper mode though iirc.

Looks like it should be possible to adjust on the `AvatarAudioOutputManager` Located on your `Head Proxy`. Although I think if a world provides configs they'd overwrite any on the avatar.

**Comment by The-furf-of-July** [comment]
I'm not sure making this a default feature would be best, but adding things that make facing attenuation systems easier to implement would be good.

One potential issue with letting users control their hearing falloff is that we need to make sure they aren't able to adjust it past what the world's settings should allow them to hear. If someone enhances their ranges too much it could cause privacy concerns.

**Comment by Pinkyyyy** [comment]
Could someone make this into an item? So you grab earmuffs and put it on your head, and they attach to your avatar, similar to how Mute Helper works, and you can also just grab them to take them off (maybe might be better to do this action through the context menu).

It would also be a visual indication to other people that hey: maybe yall should separate out a bit.

I find it strange how people even in VR decide to all just stand in 1 big group and have 5 different conversations in mini groups. Instead of you know...just walking to a different place to talk.

**Comment by MargoSizemore** [comment]
> Could someone make this into an item? So you grab earmuffs and put it on your head, and they attach to your avatar, similar to how Mute Helper works, and you can also just grab them to take them off (maybe might be better to do this action through the context menu).
> 
> It would also be a visual indication to other people that hey: maybe yall should separate out a bit.
> 
> I find it strange how people even in VR decide to all just stand in 1 big group and have 5 different conversations in mini groups. Instead of you know...just walking to a different place to talk.

Can someone let me know if this becomes a thing?

**Comment by MegaShinySnivy** [comment]
> Could someone make this into an item? So you grab earmuffs and put it on your head, and they attach to your avatar, similar to how Mute Helper works, and you can also just grab them to take them off (maybe might be better to do this action through the context menu).
> 
> It would also be a visual indication to other people that hey: maybe yall should separate out a bit.
> 
> I find it strange how people even in VR decide to all just stand in 1 big group and have 5 different conversations in mini groups. Instead of you know...just walking to a different place to talk.

My issue is that even in 'small' groups I run into issues where the audio falloff doesn't seem to be enough, and I still hear players at full volume even on the other side of the world. As someone with an audio processing disorder, this is mildly irritating and disorienting at best and forces me to get up and leave at worst

**Comment by Pinkyyyy** [comment]
I feel you, the audio falloff seems very large, i wish the normal speaking one was tighter. I also seem to get overwhelmed when multiple people are talking at the same time:  could be like 2 small separate groups, but i can still hear them slightly.

I have to constantly go into the session settings and turn people down

**Comment by akiRAM3** [comment]
I really want this. 
This is necessary for accessibility in my opinion.
While some people may experience a strong cocktail party effect, for me, that effect is weak.

**Comment by epicEaston197** [comment]
The main issue with whisper mode is that no one outside your bubble can hear you also whisper mode would be used in different circumstances and scenarios

**Comment by Catboy-Slim** [comment]
This issue definitely deserves the accessibility label. Whisper bubble doesn't really work for me because I can't change the distance at which I can hear other people, I can only sort of change my own voice broadcast distance. That, and the big bubble is very intrusive. Related issue #463

**Comment by Frooxius** [comment]
I'm trying to distill what would be the actual behaviors / features to implement from this.

I'm not familiar with the actual Earmuff mode and what it does on technical level, so descriptions/documentation of that would help.

However it sounds like best approach might be:

- Allow setting distance falloff distance & transition and how much will the voice attenuate beyond the distance
- Set "looking cone" angle and volume, so voices that are behind the user will be attenuated

Are these things that user is expected to tune constantly depending on the situation/world? Or is it more something you just set once and forget?

**Comment by Hayden-Fluff** [comment]
Another thing that could be important for earmuff mode is to have an option for it to also affect non-user sounds (anything not categorized under Voice).
I've come across many cases where incorrectly categorized sounds are just too much, or I just don't want to hear sounds that I don't care about or aren't important.

Having the ability to quickly turn this on as well would be very helpful, in VRChat you can double-click the Volume menu button in the hand menu to quickly toggle it, and I've seen that help in many cases. I'd recommend a facet that's present on the default dash to toggle this.

The lack of a feature like this is currently one of the main things keeping me from the platform, as I actively use it all the time in VRChat and consider it a very important accessibility feature for social VR, especially for someone that gets very easily overstimulated like myself.

**Comment by Frooxius** [comment]
@shiftyscales Thank you, this helps quite a bit! I think I got a better idea of things now. 

It does feel like the cone itself should go fairly to the side too - like it's not really fully a cone, but more like a parabola/dome(ish) shape?

To me the actual cone doesn't make fully sense with how hearing works - e.g. if I have difficulty hearing something, I turn my ear towards it, but with pure cone that would decrease the volume, instead of increasing it.

@Hayden-Fluff I don't think we can affect non-user sounds super well, as there's not a good way to instrument all of these. Are people often categorizing sound effects and other things as "Voice"? You should be able to turn SFX, Multimedia and such down already.

For the quick access - the new Settings UI has a system that will let you take any settings and place them in your dash or other places, so you'll be able to do that implicitly once that is done.

We probably won't place it there by default, because we want to keep the Home screen clean and we found in the past that too many options there overwhelm new users.

**Comment by JackTheFoxOtter** [comment]
I like the idea of an earmuff mode, especially a highly configurable one, however, it might be good at least taking a mental note of #463 to make sure the earmuff mode is not gonna be "overcorrecting" for the already inaccurate voice falloff. I think for a lot of people wanting an earmuff mode might be wanting a solution for the symptom of the inaccurate voice falloff. I do experience issues in crowded sessions frequently of having trouble focusing on the person in front of me, so I think I'd probably occasionally turn this on in any case.

**Comment by Frooxius** [comment]
Hmm this makes this a bit more complicated, I do wonder if we need to tackle the voice falloff first before this. 

But it sounds to me like it can be done either way. If this "overcorrects", people will just need to re-tune the earmuff settings.

**Comment by JackTheFoxOtter** [comment]
> But it sounds to me like it can be done either way. If this "overcorrects", people will just need to re-tune the earmuff settings.

Yeah that's my thinking as well. It would be useful in either case! I think both are valuable, and it will likely also be a big preference thing for different people. I just wanted to bring it up because the issues are very related!

**Comment by Hayden-Fluff** [comment]
This is less related to earmuff mode in particular, but is related to a point Shifty made in an earlier comment.

Another problem I have with only Resonite is I'm always too quiet, I don't use normalization, so to counteract this I have to add an extra 12dB of gain to my pre-processing chain to be at a "normal" volume.

As stated, Resonite is literally the only place I have this issue, and I don't feel like normalization should have to be a requirement because there are cases where I do want the dynamic range in my voice.

**Comment by FlameSoulis** [comment]
I have to agree that the larger relation of #463 is a major culprit in why this kind of feature is desired. Honestly, outside of HiFi's handling of audio, virtually all platforms seem to have an issue balancing voice volumes, and it ends up becoming a complete mess and a repeat of "the loudness war."

I was meaning to compose it, but perhaps it's more related to #73, where my Quest Pro's audio is unusually blown out, even with normalization on, whereas my Index isn't. I'm not an audio engineer, but I try to keep all microphones at equal levels if I can help it, and now seeing that may explain some oddities I've been having.

Overall, I really do suggest looking up some of the raw videos from High FIdelity's stress tests because a room full of 400+ people, all talking, and the simple fact that you could still understand people right next to you is proof that balance is possible. The concerns of this feature are what I'd view as accessibility for those who need less background noise and more clarity for those nearby, but as it stands, this applies moreso for many more people simply because the audio handling is kind of broken.

**Comment by WayaShinzui** [comment]
I will cry big fat happy tears if earmuff gets added and/or voice falloff is fixed. I frequently have to jump away and throw up my whisper bubble when there's a ton of people around and no user-made culling system to help. The current voice falloff is _painful._

The way Shifty described the earmuff settings from VRC is spot-on and that shit is so amazing. You can actually think in a room full of people.

**Comment by Frooxius** [comment]
@Hayden-Fluff We can potentially add "input gain" setting in-game, but you can also just adjust the audio input volume in your system.

But generally if you disable normalization, Resonite will just use whatever audio volume your microphone gives it. If you make that choice, that's how it'll behave and it's up to you to manage your volume level.

However this is getting a bit too off topic for this, so if you have issues there, please make a separate issue.

------------------------

I will be implementing this very soon (sometime next week), so this feature will come, I'm just trying to better understand the issue now, which is why I'm asking all the questions on this.

**Comment by HamoCorp** [comment]
Related to what was said about non-user sounds. Things like mute helpers are categorised as multi-media wouldn't work well with the mode. Also when people are having conversations in different parts of the world and have sounds on their avatar or spawned objects. Videos spawned related to a conversation usually result in everyone else going over to turn down the video player. If it's possible or in the future. having options to include other spatial sounds might be worth some consideration.

**Comment by Frooxius** [comment]
I think the solution for that would be to mark the mute helpers to Voice potentially.

But generally other sounds should be controlled by the Sound Effect and Multimedia volume sliders - if those are too loud, you can decrease their volume that way and keep Voice higher.

Doing earmuff on every single sound in the scene is something we're unlikely to do, because it adds a lot of complications:
- The earmuff will involve a number of complex calculations, so this would end up having potentially significant performance cost
- Sounds in the scene can be arbitrary and controlled in arbitrary manners - this makes it really difficult to reconcile with the earmuff options
- While we control the user voice output, we don't have this level of control for any other sounds, which might be driven by ProtoFlux and so on, which would make the behavior inconsistent

I think for all other sounds, using the global volume sliders is probably best option here. You don't need to go to the video player to turn it down - you can just do so using the Multimedia slider.

**Comment by ultrawidegamer** [comment]
Personally, I sometimes get super overwhelmed when a lot of people are talking nearby. When I played VRC, I had earmuff mode enabled 100% of the time. The key to making this feature work well is allowing configuration for multiple different scenarios. The most important settings to have are Distance, Falloff, Audio Type Switch, and Visual Indicator Toggle.

Distance: The distance before audio starts to falloff.
Falloff: The distance it takes to go from the current volume to no volume.
Audio Type Switch: A way to switch between different types of audio to ignore outside of the zone (voices, multimedia, all).
Visual Indicator Toggle: A toggle to show and hide the distance and falloff circles.

I'm super excited and can't wait for this feature to be added, as it will be a lifesaver in many of my (and I'm sure many others) social interactions.

**Comment by Frooxius** [comment]
We should keep this thread more on topic please, because it's getting a bit more unwieldy with other related issues being brought up.

If there's another issue, let's make it a separate issue please.

**Comment by ultrawidegamer** [comment]
Personally, I sometimes get super overwhelmed when a lot of people are talking nearby. When I played VRC, I had this feature enabled 100% of the time. The key to making this feature work well is allowing configuration for multiple scenarios. The most important settings to have are Distance, Falloff, Audio Type Switch, and Visual Indicator Toggle.

Distance: The distance before audio starts to falloff.
Falloff: The distance it takes to go from the current volume to no volume.
Audio Type Switch: A way to switch between different types of audio to ignore outside of the zone (voices, multimedia, all).
Visual Indicator Toggle: A toggle to show and hide the distance and falloff circles.
I'm super excited and can't wait for this feature to be added, as it will be a lifesaver in many of my (and I'm sure many others') social scenarios.

**Comment by epicEaston197** [comment]
I do feel like that this setting should be placed in its own category called "Accessibility"

**Comment by Hayden-Fluff** [comment]
To add onto my comment earlier about earmuff mode affecting sounds other than users, I say this because the earmuff mode in VRChat has an option to also affect Avatar audio (ie any audio that's on a user's avatar) and this is very helpful when in open or crowded worlds.

I understand that Resonite doesn't really have this kind of separation with audio sources, but something in this direction would be very helpful. I'm unsure if it'd be possible to only affect audio sources under avatars, but this would be a very welcome thing if it is possible.

**Comment by FlameSoulis** [comment]
I mean, it's still doable? The earmuff mode would just need to check if the audio source is parented under a user, much like how an avatar is. If it is, then we know the audio source is from a user.

However, for objects spawned by a user and then unparented, this would fall under the radar of detection.

**Comment by Frooxius** [comment]
Is there a reason that lowering the volume of sound effects, multimedia and UI sounds is not sufficient for this?

Using those volume sliders you should be able to lower those sounds so they're quieter than voice.

**Comment by epicEaston197** [comment]
> Is there a reason that lowering the volume of sound effects, multimedia and UI sounds is not sufficient for this?
> 
> Using those volume sliders you should be able to lower those sounds so they're quieter than voice.

I can understand this if you're in a busy Club World you still want to be able to hear the music

**Comment by Frooxius** [comment]
> > Is there a reason that lowering the volume of sound effects, multimedia and UI sounds is not sufficient for this?
> > Using those volume sliders you should be able to lower those sounds so they're quieter than voice.
> 
> I can understand this if you're in a busy Club World you still want to be able to hear the music

This wouldn't help you. Either the Ear Muff would make it quieter, or the Multimedia slider would. In either case, if you turn things quieter through some feature, they'll be quieter.

**Comment by epicEaston197** [comment]
Yeah that's exactly what I was saying I wasn't against that you want players voices to be quieter within a certain range but you still wanna be able to hear the environment

**Comment by epicEaston197** [comment]
> I'd love to earmuff mode to easily switchable. I personally do not use this all the time, but I want activate when I'm in crowded areas. Therefore, I hope it can be easily switched on and off.
> 
> For example, it would be great if it were as accessible as switching the 'sitting mode' from the dash menu.
> 
> By the way, I really appreciate you considering implementing this feature.

This would likely be a setting somewhere and there's a feature that's planned but it will allow you to place facets onto your hands so although it might not be quick to access in your settings menu now, if it's added it will be way easier to access in the future when it's on your avatar's hands

**Comment by Frooxius** [comment]
Earmuff settings have been now added in 2024.5.1.68, thanks for requesting this everyone, I hope it helps!

I'm closing this issue, since the initial version is done, but this doesn't mean we can't iterate on this more!

If something doesn't work quite right or you need more controls and such, please feel free to make new issues! But give it a spin first, play with the settings, see how well it works.
